### ISSUE TITLE
Initial shader compilation on first run is slow (causes missed frames on first launch)

### ISSUE THREAD

**Issue by sethladd** [flutter/flutter#5363]
On Android, open Gallery, tap into Shrine. Then, tap an item. Notice how the initial animation up misses frames. Then, tap back. Notice more missed frames. Then, tap on the same item. The animation up is smooth.

Seems like something is happening on the first animation for hero'ing an item up. Not sure what, but it's causing us to miss frames.

After the first animation, things are smooth, so we can get to great animations with the items.

**Comment by chinmaygarde** [comment]
Does the stutter happen on tapping another item? If it does, then bottleneck is the population of the raster cache entry. If it doesn't, its probably the setup of the shader.

**Comment by sethladd** [comment]
It appears that the missing frames is visible only on the first hero of the item.

Order:

* tap sunglasses
* tap back
* tap radio

By the time I get to radio, things _feel_ faster.

Looking at the perf overlay, the radio is taking less time to hero up than the sunglasses did. It's still missing frames, but not as severely.

The radio doesn't seem to be missing the first few frames like the sunglasses does. I don't see the first initial "big" pause with the radio, like I do the sunglasses.

**Comment by sethladd** [comment]
Did anyone take a look at this one? Curious what we learned. There's something happening initially in Shrine that makes it jank, but then it's all good. Compiling shaders? Can we mitigate that in the future?

**Comment by sethladd** [comment]
This is definitely more obvious on Android hardware.

**Comment by eseidelGoogle** [comment]
I recall a theory being tossed around the cube that initial jank in in pesto/shrine might be due to custom fonts.  We'll need to trace.

**Comment by abarth** [comment]
The custom font issue is different than this one.  The custom fonts make the initial load of pesto and shrine slow (because we need to load the fonts off disk).  This issue is likely something else.

**Comment by eseidelGoogle** [comment]
Looks like a giant 18ms build it at least part of the problem:
<img width="392" alt="screen shot 2017-02-03 at 10 14 21 am" src="https://cloud.githubusercontent.com/assets/11857803/22602807/9c7b7d1c-e9f9-11e6-922b-7276eb6757ba.png">

**Comment by eseidelGoogle** [comment]
This is the trace: https://ufile.io/a641c  Doesn't seem to load in chrome-tracing, filed: https://github.com/dart-lang/sdk/issues/28628, but might load in Observatory.

**Comment by sethladd** [comment]
Reported on Reddit: https://www.reddit.com/r/androiddev/comments/5o8v0g/anyone_using_dart_with_flutter_how_is_it/ddja6uw/

**Comment by eseidelGoogle** [comment]
Based on the above this should be at least M3. :)

**Comment by eseidelGoogle** [comment]
This is presumably very related to https://github.com/flutter/flutter/issues/4991 ?

**Comment by ScottPierce** [comment]
This may be related to https://github.com/flutter/flutter/issues/6920

**Comment by eseidelGoogle** [comment]
@lukef just reported that animations fail on first navigation but succeed on subsequent navigations and that such is an important blocker for posse.  He may have been hitting this bug?

**Comment by chinmaygarde** [comment]
I collected [traces](https://github.com/flutter/flutter/files/984490/BeforeAfter.zip) for both the first run and subsequent runs of entering the shrine demo in the gallery.

There are two things that are different about the first run that drops the frames (both these issues are absent on subsequent runs):
* Right after the first frame is invoked, there is a `Dart_InvokeClosure` that blocks for an extremely long time (~80ms). This causes subsequent frames to be blocked on the producer on the UI thread (i.e, not the consumer on the GPU). We need to figure out what is being done here and why this happens only on the first run.
* The images on the first page need to be decoded. This happens only on the first run but should not be an issues as the decoding is asynchronous. However, I would want to verify that neither the producer nor the consumer is doing a wait on resolution of that image.

![screen shot 2017-05-08 at 11 32 36 am](https://cloud.githubusercontent.com/assets/44085/25819519/10d91fce-33e3-11e7-9941-95c2a0115309.png)

**Comment by chinmaygarde** [comment]
After chatting with the team, the issue seems to be that even though the decompression and upload of the image to the GPU is asynchronous, the handoff of the compressed image data from the platform happens on the UI thread. There is no need for this as we can make the IO thread itself wait for the source data.

**Comment by chinmaygarde** [comment]
To illustrate the issue further, I added a trace in the `ZipAssetBuffer::GetAsBuffer` routine. That does seem to be the bottleneck. We can move that entire block to the IO thread.
![screen shot 2017-05-08 at 12 52 45 pm](https://cloud.githubusercontent.com/assets/44085/25822550/c727f8c2-33ed-11e7-95c0-72d5ec546725.png)

**Comment by eseidelGoogle** [comment]
Posse has hit this and this will block them.  Adding to customer blockers for May.

**Comment by sethladd** [comment]
Do we have their repo case (curious if it's the same thing), or this is just a high-level concern?

**Comment by lukef** [comment]
I would say most of the gallery is affected by some performance issue and my suspicion is that its related to this (but i could be WILDLY wrong). 

In a different app, the first time you try to use `Navigate` it will first stutter and then kill the animation and it's super disorienting.

**Comment by chinmaygarde** [comment]
[This patch](https://github.com/flutter/engine/pull/3682) has landed and it address the slow ZipAssetStore access. @jason-simmons has also said that he will work on removing the FLX files completely so it should be faster still.

Please re-open if the performance is still not satisfactory. Images that are extremely large should still cause stalls (but without any spikes showing up in the performance overlay). This can be fixed by either using smaller images or throwing up placeholders or spinners till the image is fully loaded. An example of such an issue is https://github.com/flutter/flutter/issues/9972.

**Comment by oliverjanik** [comment]
When will this land int the Gallery app?

**Comment by sethladd** [comment]
Reopening because we're still missing frames on initial animation.

**Comment by Hixie** [comment]
cc @eseidel We don't have anyone to work on this right now.

**Comment by eseidelGoogle** [comment]
Based on @chinmaygarde's comments above, this sounds like it my fall out of @szakarias's work on flx.

**Comment by Hixie** [comment]
This is probably a specific case of https://github.com/flutter/flutter/issues/813.

**Comment by cbracken** [comment]
De-duped with #6920, which was specific to Pesto. Maintaining this as the catch-all for the Gallery first-run jank issue, which as @Hixie points out, is a specific case of #813.

**Comment by newfivefour** [comment]
This is also reported on the top comment on Google Play for the Flutter.io Gallery app.

**Comment by mit-mit** [comment]
I still see this in the most recent Gallary build from Testflight.

**Comment by xster** [comment]
Breaking this down into some sub-bugs #14028, #17270, #17271, #13736

**Comment by cbracken** [comment]
Assigning @chinmaygarde as he's working on #14028 which is the main remaining open sub-bug for this meta bug.

**Comment by chinmaygarde** [comment]
https://github.com/flutter/flutter/issues/14028 and a couple of follow-ups have landed.

**Comment by Hixie** [comment]
I intend to look at this in the coming weeks to see where we are at with older phones.

**Comment by liyuqian** [comment]
@bsalomon : please let us know as soon as Ethan added the ability to cache the GLSL :-)

**Comment by liyuqian** [comment]
Duplicate of https://github.com/flutter/flutter/issues/813

**Comment by goderbauer** [comment]
Closed as duplicate of #813.
